### PR TITLE
[GH-9807] Fix: initialize potentially empty collections at the hydration complete

### LINKED ISSUE #9807: Collection is hydrated as empty if first row contains nulls from a leftJoin
### Bug Report

|    Q        |   A
|------------ | ------
| BC Break    | no
| Version     | 2.11.3

#### Summary

When the first row of a leftJoin result contains null values (because of a SqlFilter), the collection is hydrated as empty.

#### Current behavior

The collection is hydrated as empty if the first row is nullish.

#### How to reproduce

Details for reproduction is in https://github.com/freiondrej-lmc/doctrine_orm_bug_reproduction .

#### Expected behavior

The Collection should be hydrated with existing data.

### PR DESCRIPTION
**Pull request by popov-a-e**
Closes https://github.com/doctrine/orm/issues/9807

### CODE REVIEW

**Comment by popov-a-e** [comment]
Anybody alive here? 🙂

File: tests/Doctrine/Tests/ORM/Functional/Ticket/GH9807Test.php [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $this->_schemaTool->createSchema(
+            [
+                $this->_em->getClassMetadata(GH9807Main::class),
+                $this->_em->getClassMetadata(GH9807Join::class),
+            ]
+        );
```
Comment (greg0ire) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/GH9807Test.php]: Please use `OrmFunctionalTestCase::createSchemaForModels` instead.

File: tests/Doctrine/Tests/ORM/Functional/Ticket/GH9807Test.php [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+    /**
+     * @ORM\ManyToMany(targetEntity="GH9807Join", inversedBy="starts")
+     *
+     * @var Collection<int, GH9807Join>
+     */
+    private $joins;
+
+    /**
+     * @return Collection
```
Comment (greg0ire) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/GH9807Test.php]:
```suggestion
     * @return Collection<int, GH9807Join>
```

File: tests/Doctrine/Tests/ORM/Functional/Ticket/GH9807Test.php [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+ */
+class GH9807Main
+{
+    /**
+     * @var int
+     * @Column(type="integer")
+     * @Id
+     * @GeneratedValue
+     */
+    protected $id;
```
Comment (greg0ire) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/GH9807Test.php]: Why are ids protected in both entities?

**Comment by greg0ire** [comment]
Please improve your commit message according to [the contributing guide](https://www.doctrine-project.org/contribute/index.html#working-on-topic-branches).

**Comment by popov-a-e** [comment]
Thanks for the comments @greg0ire, I've fixed it

File: lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php [hunk 4]
```diff
@@ -44,6 +44,9 @@ class ObjectHydrator extends AbstractHydrator
     /** @var mixed[] */
     private $initializedCollections = [];
 
+    /** @var mixed[] */
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php]: This is correct but I think you can be more precise.

```suggestion
    /** @var array<string, PersistentCollection> */
```

File: lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php [hunk 5]
```diff
@@ -148,6 +151,12 @@ protected function hydrateAllData()
             $coll->takeSnapshot();
         }
 
+        foreach ($this->uninitializedCollections as $coll) {
+            if (! $coll->isInitialized()) {
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php:157]: > the collection will be considered "existed", and "existed" collections are only replaced if REFRESH hint is present

Did you mean to write "initialized"?

**Review by greg0ire — COMMENTED** [review]
I forgot to mention that in the contribution guidelines, but can you wrap the body of your commit message (which is great, good job!) at 72 chars?

File: lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php (hunk 4 above)
Comment (popov-a-e) [reply lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php]: I thought about this ofc but didn't want to differ from the same other collections. This file is to refactor anyway, aint' it?

File: lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php (hunk 5 above)
Comment (popov-a-e) [reply lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php:157]: No, there is such a logic that if a Collection is PersistentCollection and initialized it is added to "existing"
```
else {
            // Is already PersistentCollection, and DON'T REFRESH or FETCH-JOIN!
            $this->existingCollections[$oid . $fieldName] = $value;
 ```

File: lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php (hunk 4 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php]: Yes, but I'm likely to be the one doing it. Please don't give me more work :pray:

Comment (popov-a-e) [reply lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php]: Huh, OK, no problem

File: lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php (hunk 5 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php:157]: I see 👍; please use "existing" then.

**Comment by popov-a-e** [comment]
> I forgot to mention that in the contribution guidelines, but can you wrap the body of your commit message (which is great, good job!) at 72 chars?

Hm, do you mean leaving only the headline? 72 chars is not enough to unfold the reasoning

**Comment by greg0ire** [comment]
> Hm, do you mean leaving only the headline? 

No, I mean wrapping as in adding a linebreak when a line is 72 chars long. With vim properly configured, you can do that with `gqq`

**Review by greg0ire — DISMISSED** [review]

**Comment by popov-a-e** [comment]
Got it, thanks for the hint. I misunderstood "wrapping" for "reducing"
I use PHPStorm however. Consider me casual developer for not using vim :D

**Review by SenseException — DISMISSED** [review]

File: lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php [hunk 6]
```diff
@@ -148,6 +151,12 @@ protected function hydrateAllData()
             $coll->takeSnapshot();
         }
 
+        foreach ($this->uninitializedCollections as $coll) {
```
Comment (beberlei) [inline lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php:156]: I believe you need to reset this in cleanup()

**Review by beberlei — COMMENTED** [review]
Good from my first review

File: lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php (hunk 6 above)
Comment (popov-a-e) [reply lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php:156]: Indeed, makes sense

Comment (SenseException) [reply lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php:156]: Shouldn't there be a test for the cleanup of the new property then?

Comment (popov-a-e) [reply lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php:156]: I believe that the details of the implementation are not something that needs to be tested by itself. There shouldn't be tests that break for on a correctly working class with an unchanged interface. Like there are no tests for other collections or private fields that are getting reset, for example `$existingCollections` or `$initializedCollections`.

Comment (popov-a-e) [reply lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php:156]: Tests for checking if collections are getting initialized or not in different circumstances are nice to have but those are not related to the original issue.

**Review by SenseException — APPROVED** [review]

**Comment by greg0ire** [comment]
Thanks @popov-a-e !

**Pull request merged**